uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-details-tab.component.js

```diff
@@ -162,31 +158,58 @@ define([
     }());
 
     /**
-     * Calculate change in balance, it updates local balance variables.
-     *
-     * @return {Promise} empty promise if all required params are not set
-     *   otherwise promise from server
+     * Checks To Date and flushes it if it is the same or before
+     * the "from date" in case of the multiple day request.
      */
-    function calculateBalanceChange () {
-      vm._setDateAndTypes();
-      vm._toggleBalance();
+    function checkToDate () {
+      if (
+        vm.uiOptions.toDate && vm.uiOptions.fromDate && vm.uiOptions.multipleDays &&
+        moment(vm.uiOptions.toDate).isSameOrBefore(vm.uiOptions.fromDate)
+      ) {
+        vm.uiOptions.toDate = null;
+      }
+    }
 
-      if (!vm._canCalculateChange()) { return $q.resolve(); }
+    /**
+     * Calculates closing balance which is opening balance minus change amount.
+     * We use "+" operation since the change amount is negative.
+     */
+    function calculateClosingBalance () {
+      vm.balance.closing = vm.balance.opening + vm.balance.change.amount;
+    }
 
-      vm.loading.showBalanceChange = true;
+    /**
+     * Converts given date to javascript date as expected by uib-datepicker
+     *
+     * @param {String} date from server
+     * @return {Date}
+     */
+    function convertDateFormatFromServer (date) {
+      return moment(date, sharedSettings.serverDateFormat).toDate();
+    }
 
-      return vm.request.calculateBalanceChange(vm.selectedAbsenceType.calculation_unit_name)
-        .then(setBalanceChange)
-        .catch(handleError);
+    /**
+     * Converts given date to server format
+     *
+     * @param {Date} date
+     * @return {String} date converted to server format
+     */
+    function convertDateToServerFormat (date) {
+      return moment(date).format(sharedSettings.serverDateFormat);
     }
 
     /**
-     * Change handler when changing no. of days like Multiple Days or Single Day.
-     * It will reset dates, day types, change balance.
+     * Handles the change of days selection mode (single day, multiple days)
+     *
+     * @return {Promise}
      */
-    function changeInNoOfDays () {
-      vm._reset();
-      vm._calculateOpeningAndClosingBalance();
+    function daysSelectionModeChangeHandler () {
+      checkToDate();
+      resetUIDayTypesTimeAndDeductions('to');
+      setRequestDates();
+
+      return performBalanceChangeCalculation()
+        .then(vm.setDaysSelectionModeExtended);
```

Comment:
why should this happen only after the balance change had been calculated? I think it should happen first, otherwise it means that the custom logic depends on whether the balance change calculation has been successful

Reply:
Fixed and wrote a test case, but the problem is that it already succeeds even without the fix due to the handleError function that prevents balance calculation to reject. If I return `$q.reject()` in `handleError()`, then test fails without the fix and succeeds with the fix.